test/boost/secondary_index_test.cc

```diff
@@ -1560,6 +1560,305 @@ SEASTAR_TEST_CASE(test_token_order) {
+                { int32_type->decompose(3), int32_type->decompose(2), int32_type->decompose(2), int32_type->decompose(1), int32_type->decompose(1) }
+            });
+
+            msg = cquery_nofail(e, "SELECT pk1, pk2, ck1, ck2, v2 FROM t WHERE token(pk1, pk2) >= token(1, 1) AND token(pk1, pk2) <= token(3, 2) AND v2 = 1 AND ck1 = 1 ALLOW FILTERING");
+            assert_that(msg).is_rows().with_rows({
+                { int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1) },
+                { int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(3), int32_type->decompose(1) },
+                { int32_type->decompose(3), int32_type->decompose(2), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1) },
+                { int32_type->decompose(3), int32_type->decompose(2), int32_type->decompose(1), int32_type->decompose(3), int32_type->decompose(1) },
+            });
+
+            msg = cquery_nofail(e, "SELECT pk1, pk2, ck1, ck2, v2 FROM t WHERE token(pk1, pk2) >= token(1, 1) AND token(pk1, pk2) <= token(3, 2) AND v2 = 1 AND ck1 = 1 AND ck2 = 1 ALLOW FILTERING");
+            assert_that(msg).is_rows().with_rows({
+                { int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1) },
+                { int32_type->decompose(3), int32_type->decompose(2), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1) },
+            });
+
+            msg = cquery_nofail(e, "SELECT pk1, pk2, ck1, ck2, v2 FROM t WHERE token(pk1, pk2) >= token(1, 1) AND token(pk1, pk2) <= token(3, 2) AND v2 = 1 AND ck2 = 1 ALLOW FILTERING");
+            assert_that(msg).is_rows().with_rows({
+                { int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1) },
+                { int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(2), int32_type->decompose(1), int32_type->decompose(1) },
+                { int32_type->decompose(3), int32_type->decompose(2), int32_type->decompose(1), int32_type->decompose(1), int32_type->decompose(1) },
+                { int32_type->decompose(3), int32_type->decompose(2), int32_type->decompose(2), int32_type->decompose(1), int32_type->decompose(1) }
+            });
+        };
+
+        // Do tests for each column (excluding partition key columns - cannot mix partition column 
+        // restrictions and TOKEN restrictions). First run the query without index, then with global index,
+        // and finally with local index.
+        eventually(do_tests_ck1);
+        cquery_nofail(e, "CREATE INDEX t_global ON t(ck1)");
+        eventually(do_tests_ck1);
```

Comment:
Houston, I think we have a problem...
Earlier I noted that I don't like to use eventually() for these tests. But in this case, I think using eventually isn't just annoying for development - it may even be wrong... The thing is (but please correct me if I'm wrong!), that this test will succeed even **without** an index - as you checked above. So if the index was not yet done creating, the test may simply pass without the index, and you check won't check anything!
So I think you really, as I suggested above, need to wait until the index is fully available (e.g., fetch all the different index-column values and see you can read the entire table using the index), and only then run the test.

Reply:
From what I discussed with @psarna, my understanding is that after `CREATE INDEX`, the schema is updated (before `do_tests_ck1` is performed). Therefore next `do_tests_ck1` will read from the index (not without) - a few initial tries will fail as the index is not fully built (therefore `eventually`). Experimentally, if I did `do_tests_ck1` without `eventually` after `CREATE INDEX` it failed (not observed the query being done without index).

@psarna Is my understanding correct?

Reply:
It is. Here's the relevant code:
https://github.com/scylladb/scylla/blob/f55b522c1bdc25dbae56c1e1ff6fe1c14c9ecc60/service/migration_manager.cc#L958-L959

Reply:
I see. Ok.